dropshot/src/from_map.rs

```diff
@@ -46,22 +59,68 @@ impl MapValue for String {
     }
 }
 
+// To handle headers, which are case-insensitive, we use a trait to adapt
+// deserialization of structs to be case-insensitive.
+
+trait Casing {
+    fn rename(fields: &'static [&'static str]) -> BTreeMap<String, String>;
+}
+
+enum CaseSensitive {}
+enum CaseInsensitive {}
+
+impl Casing for CaseSensitive {
+    fn rename(_fields: &'static [&'static str]) -> BTreeMap<String, String> {
+        Default::default()
+    }
+}
+impl Casing for CaseInsensitive {
+    fn rename(fields: &'static [&'static str]) -> BTreeMap<String, String> {
+        fields
+            .iter()
+            .map(|field| (field.to_lowercase(), field.to_string()))
```

Comment:
Is the lifetime too much of a pain to return the "value" here as a `&'static str`? seems unnecessary to perform the string allocation (via `field.to_string()`) on the case-unchanged string here